src/data_model/eng2math.jl

```diff
@@ -0,0 +1,811 @@
+
+        if kron_reduced
+            filter = _kron_reduce_branch!(math_obj,
+                Vector{String}([]), ["gs", "bs"],
+                eng_obj["connections"], kr_neutral
+            )
+            connections = eng_obj["connections"][filter]
+            _pad_properties!(math_obj, ["gs", "bs"], connections, kr_phases)
+        else
+            math_obj["connections"] = eng_obj["connections"]
+        end
+
+        data_math["shunt"]["$(math_obj["index"])"] = math_obj
+
+        push!(data_math["map"], Dict{String,Any}(
+            "from" => name,
+            "to" => "shunt.$(math_obj["index"])",
+            "unmap_function" => "_map_math2eng_shunt!",
+        ))
+    end
+end
+
+
+"converts engineering load components into mathematical load components"
+function _map_eng2math_load!(data_math::Dict{String,<:Any}, data_eng::Dict{<:Any,<:Any}; kron_reduced::Bool=true, kr_phases::Vector{Int}=[1, 2, 3], kr_neutral::Int=4)
+    for (name, eng_obj) in get(data_eng, "load", Dict{Any,Dict{String,Any}}())
+        math_obj = _init_math_obj("load", name, eng_obj, length(data_math["load"])+1)
+
+        connections = eng_obj["connections"]
+
+        math_obj["load_bus"] = data_math["bus_lookup"][eng_obj["bus"]]
+
+        math_obj["pd"] = eng_obj["pd_nom"]
+        math_obj["qd"] = eng_obj["qd_nom"]
+
+        if kron_reduced
+            if math_obj["configuration"]==WYE
+                @assert connections[end]==kr_neutral "for wye-connected loads, if kron_reduced the connections list should end with a neutral"
+                _pad_properties!(math_obj, ["pd", "qd"], connections[connections.!=kr_neutral], kr_phases)
+            else
+                _pad_properties_delta!(math_obj, ["pd", "qd"], connections, kr_phases)
+            end
+        else
+            math_obj["connections"] = connections
+        end
+
+        math_obj["vnom_kv"] = eng_obj["vm_nom"]
+
+        data_math["load"]["$(math_obj["index"])"] = math_obj
+
+        push!(data_math["map"], Dict{String,Any}(
+            "from" => name,
+            "to" => "load.$(math_obj["index"])",
+            "unmap_function" => "_map_math2eng_load!",
+        ))
+   end
+end
+
+
+"converts engineering generators into mathematical generators"
+function _map_eng2math_generator!(data_math::Dict{String,<:Any}, data_eng::Dict{<:Any,<:Any}; kron_reduced::Bool=true, kr_phases::Vector{Int}=[1, 2, 3], kr_neutral::Int=4)
+    for (name, eng_obj) in get(data_eng, "generator", Dict{String,Any}())
+        math_obj = _init_math_obj("generator", name, eng_obj, length(data_math["gen"])+1)
+
+        connections = eng_obj["connections"]
+        nconductors = data_math["conductors"]
+
+        math_obj["gen_bus"] = data_math["bus_lookup"][eng_obj["bus"]]
+        math_obj["gen_status"] = Int(eng_obj["status"])
+        math_obj["control_mode"] = get(eng_obj, "control_mode", DROOP)
+        math_obj["pmax"] = get(eng_obj, "pg_ub", fill(Inf, nconductors))
```

Comment:
why is `pg_ub` treated differently here than the other bounds?

Reply:
pmax is an expected value by PowerModels and will break if not included